python/cudf/cudf/core/frame.py

```diff
@@ -236,9 +237,24 @@ def _concat(cls, objs, axis=0, ignore_index=False, sort=False):
                     result_index_length += len(obj)
                     empty_has_index = empty_has_index or len(obj) > 0
 
-        # Get a list of the unique table column names
-        names = [name for f in objs for name in f._column_names]
-        names = OrderedDict.fromkeys(names).keys()
+        if join == "inner":
+            old_index = []
+            for obj in objs:
+                index = obj.columns
+                old_index.append(index)
+            index = old_index[0]
+            for other in old_index[1:]:
+                index = index.intersection(other)
+            names = list(index)
+            for obj in objs:
+                for col in obj.columns:
+                    if col not in names:
+                        obj = obj.drop(columns=[col])
```

Comment:
I think it might incur a little less overhead if we compute the complete list of column names to be dropped for each object and then pass them as a list to `drop`. In the end it's a loop anyways, but I suspect it's faster to call `drop` less times overall.

Reply:
Yeah, I've been trying a couple of options for this. I'll keep this in mind and commit a new option once it's passed all the tests.

Reply:
It will reduce loop complexity. But why will it call `drop` less time?

Reply:
Because right now there's a separate call to `drop` for each of the columns that are not found in `names`. Instead we could do something like 

```
cols_not_in_names = set(names) - set(obj.columns)
obj = obj.drop(columns=cols_not_in_names)
```
That way there'd be only one call to `drop` per object.

Reply:
I decided to use a list comprehension to reduce the loop complexity, then added in your suggestion for the drop Brandon. Feel free to take a look at the updated code. Thanks :D